### DISCUSSION TITLE
Using Apache Tika with dnGrep

### DISCUSSION THREAD

**Discussion by doug24** [dnGrep/dnGrep#1049] [Show and tell]
dnGrep searches plain text files - it cannot directly search binary file formats like PDF, Word, and Excel.  That's where plugins come in - plugins extract plain text from the document files into the format that dnGrep can search.

[Apache Tika](https://tika.apache.org/) toolkit detects and extracts metadata and text from over a [thousand different file types](https://tika.apache.org/3.0.0-BETA/formats.html).  Some notable formats that Tika can handle includes:
- OpenDocument (.odt, .odp)
- EBooks (.epub)
- HTML (removes formatting tags leaving plain text)
- Word (old-style word .doc files without having Word installed)
- Image, music, and video file metadata
- Mail formats

If you want to use Tika with dnGrep, follow the download instructions for a version of the **tika-app.jar** from the Tika website.  

You will also need to install a Java JRE or JDK if you do not have one already - either from java.com or any OpenJDK build. 

Next, create a .plugin file to tell dnGrep how call Tika. See the [Bring Your Own Plugin](https://github.com/dnGrep/dnGrep/wiki/Bring-Your-Own-Plugin) page in the dnGrep help for details on making a plugin file and where to put the file.  Here is an example plugin file named `dnGrep.Tika.plugin`:

```
Name=Tika
Application=java.exe
Arguments=-jar tika-app-3.0.0-BETA.jar --text "{inputFile}"
WorkingDirectory=C:\Apps\apache-tika
Extensions=odt, ods, odp, odf, epub, doc, htm, html
```

In this example, the directory for java.exe is on the Windows system path, so no path is shown here.  Alternatively, you can specify the full path to java.exe.

This example is using the tika-app-3.0.0-BETA.jar file.  The jar file is in the C:\Apps\apache-tika working directory.  

Another option would be to put the jar file in a directory named C:\Program Files\dnGrep\Plugins\Tika along with the new dnGrep.Tika.plugin file. The WorkingDirectory can be set to C:\Program Files\dnGrep\Plugins\Tika, or just leave it blank to use the default working directory (same directory where the .plugin file is located). 

The arguments tell the Tika app to extract text from the input file supplied by dnGrep when it calls Tika.  Tika will return the extracted text to standard output, where dnGrep will read it. 

### Important note about extensions and plugins 

A file extension should be associated with only one active plugin or archive, otherwise the behavior will be unpredictable.  For example, epub is one of the standard archive extensions in dnGrep - to use Tika with epub files, the epub extension must be removed from the list of Archive extensions in the dnGrep Options dialog.  Similarly, Tika can extract text from the old-style Word .doc files, but to use Tika on these files the standard Word plugin should be disabled in the dnGrep Options dialog.

**Comment by ragebear00** [comment]
the plugin works brilliantly for English, but not for CJK or unicode. The problem appears caused by the output. Even thought the ouput is corrected in chcp 65001 console, dnGREP cannot catch it correctly. 

I made the following changes, and it works great now for Unicode and CJK.

Create a tika2txt.bat in the plugin folder, such as 'C:\_a\dnGrep.4.1.92.0.x64\Plugins\TikaSearch'
```
for %%f in (%1) do set filename=%%~dpnf
set "textfile=%filename%.txt"
C:\TmpWF1\ServerWarez\OpenJDKJRE64\bin\java.exe -jar "C:\_a\dnGrep.4.1.92.0.x64\Plugins\TikaSearch\tika-app-2.9.2.jar" --text %1 > "%textfile%" 
move "%textfile%" %2
```
create  "C:\_a\dnGrep.4.1.92.0.x64\Plugins\TikaSearch\dnGREP.Tika.plugin"
```
Name=Tika
Application=cmd.exe
Arguments=/c tika2txt.bat "{inputFile}" "{outputFile}"
WorkingDirectory=C:\_a\dnGrep.4.1.92.0.x64\Plugins\TikaSearch
Extensions=odt, ods, odp, odf, epub, msg, rtf
```

To be improved: can I create the outputfile directly in the temp folder, avoid moving it?

**Reply by doug24** [reply]
Yes, instead of creating the textfile as an intermediate location, just redirect the output to the {outputFile}.  Your whole tika2text.bat file would be:
`C:\TmpWF1\ServerWarez\OpenJDKJRE64\bin\java.exe -jar "C:\_a\dnGrep.4.1.92.0.x64\Plugins\TikaSearch\tika-app-2.9.2.jar" --text %1 > %2`

The plugin file would be the same as you show.

**Comment by doug24** [comment]
The encoding problem CJK and Unicode is because dnGrep is reading the standard output from Tika using some system default encoding, whatever that is. With all the issues around text encoding, I should have learned not to accept the default for anything.  

In the next release I will modify the plugin code to copy the standard output to a binary memory stream, check the encoding, and then get the text from the stream using the detected encoding.  This should allow the original configuration to work as well. I think both options will perform about the same.

**Comment by ragebear00** [comment]
launch Tika server can avoid launch jar for each instance and improve the tika conversation substantially and reduce CPU loading.

Step 1: download Tika server and start the server.

```
java -jar "C:\dnGrep\Plugins\TikaSever\tika-server-standard-2.9.2.jar" --host=127.0.0.1 --port=9998
```

Step: create a dnGREP.TikaServer.plugin with the following

```
Name=TikaServer
Application=curl.exe
Arguments=-g -T "{inputFile}" http://localhost:9998/tika --header "Accept: text/plain"
WorkingDirectory=C:\dnGrep\Plugins\TikaSever
Extensions=odt, ods, odp, odf, epub, msg, rtf, html, htm, doc
```

**Comment by ragebear00** [comment]
Since Apache Tika support rar, zip, 7z and other compressed file extraction. You can remove those types in the Options -> Archive Files those types, and add in the Options -> Plug-ins -> Tika ->Extensions. 

In the future, the rar, zip and 7z can be treated the same as a PDF or Word file. All files in the compressed file will be converted into ONE text file. 

It can take advantage of Cache implemented. So such files will NOT be extracted again, if the zip files unchanged and hash value is the same. The speed to search those files can be significantly improved.
